xds: Add counter and gauge metrics

DNVindhya opened the pull request:
This PR implements xDS client defined in [A78](https://github.com/grpc/proposal/blob/master/A78-grpc-metrics-wrr-pf-xds.md#xdsclient).

Counters
- grpc.xds_client.server_failure
- grpc.xds_client.resource_updates_valid
- grpc.xds_client.resource_updates_invalid

Gauges
- grpc.xds_client.connected
- grpc.xds_client.resources

The `grpc.xds.authority` label is missing, and will be added in a later PR.

ejona86 reviewed (commented):
Sending what I have.

ejona86 commented on xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:
```diff
@@ -120,21 +118,22 @@ public void accept(BatchRecorder recorder) {
+          callbackMetricReporter);
+      // Normally this shouldn't take long, but adding a timeout to avoid indefinite blocking
+      Void unused = ret.get(5, TimeUnit.SECONDS);
+      SettableFuture<Void> ret1 = this.xdsClient.reportServerConnections(callbackMetricReporter);
```
Run the two concurrently (start both, and then get() both), so we don't pay any latencies twice?

DNVindhya replied:
Makes sense, started both.

ejona86 commented on xds/src/test/java/io/grpc/xds/XdsClientMetricReporterImplTest.java:
```diff
@@ -139,16 +140,19 @@ public void setXdsClient_reportCallbackMetrics_resourceCountsFails() {
 
   @Test
   public void metricGauges() {
+    XdsClientMetricReporterImpl spyReporter = spy(
```
Never, ever use a spy. The way they are implemented essentially invalidates any testing using them. Use a mock with `delegatesTo(realImpl)`.

DNVindhya replied:
When I was modifying the test to remove spy, realized I do not need `delegatesTo` either.

DNVindhya commented on xds/src/main/java/io/grpc/xds/XdsClientMetricReporter.java:
```diff
@@ -0,0 +1,93 @@
+   * Interface for reporting metrics through callback.
+   *
+   */
+  interface CallbackMetricReporter {
```
Split the interface into two different interfaces. PTAL

larry-safran commented on xds/src/test/java/io/grpc/xds/XdsClientMetricReporterImplTest.java:
```diff
@@ -110,85 +116,114 @@ public void reportServerFailure() {
         .registerBatchCallback(gaugeBatchCallbackCaptor.capture(), any(), any());
     gaugeBatchCallbackCaptor.getValue().accept(mockBatchRecorder);
+    // Verify that the xdsClient methods were called
+    verify(mockXdsClient).reportResourceCounts(any(ResourceCallback.class));
```
You don't need to put the class name in anymore (it used to be a problem well before I was on the team).  It's easier to read without it and doesn't need to be changed when the argument changes.

ejona86 commented on xds/src/main/java/io/grpc/xds/SharedXdsClientPoolProvider.java:
```diff
@@ -111,19 +113,19 @@ public ImmutableList<String> getTargets() {
+  protected class RefCountedXdsClientObjectPool implements ObjectPool<XdsClient> {
 
-    private static final ExponentialBackoffPolicy.Provider BACKOFF_POLICY_PROVIDER =
+    private final ExponentialBackoffPolicy.Provider backoffPolicyProvider =
```
This isn't providing any value as `non-static`. Either we should move it up to the provider so it can remain `static` or construct the object when creating XdsClientImpl.

(Actually, we could make Provider a singleton in ExponentialBackoffPolicy itself. But the object is small, so it doesn't matter much.)

DNVindhya replied:
Moved it up to the provider.

ejona86 commented on xds/src/main/java/io/grpc/xds/XdsNameResolver.java:
```diff
@@ -127,6 +128,7 @@ final class XdsNameResolver extends NameResolver {
   private final long randomChannelId;
 
   private volatile RoutingConfig routingConfig = RoutingConfig.empty;
+  private volatile MetricRecorder metricRecorder;
```
Why is this volatile? It is set once during construction. Seems it could even be `final`.

DNVindhya replied:
That was an oversight, updated.

ejona86 commented on xds/src/main/java/io/grpc/xds/XdsServerWrapper.java:
```diff
@@ -171,7 +172,7 @@ public void run() {
   private void internalStart() {
     try {
-      xdsClientPool = xdsClientPoolFactory.getOrCreate("");
+      xdsClientPool = xdsClientPoolFactory.getOrCreate("", new MetricRecorder() {});
```
Should there be a TODO here?

DNVindhya replied:
Added

ejona86 commented on xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java:
```diff
@@ -504,9 +533,69 @@ private <T extends ResourceUpdate> void handleResourceUpdate(
+    syncContext.execute(() -> {
+      Map<XdsResourceType<?>, Map<String, Long>> resourceCountsByType =
+          getResourceCountsByType();
+      reportResourceCountsToCallback(callback, resourceCountsByType);
```
The point of using a callback is mostly to avoid allocating the map. If you're going to construct the map, you can instead just "return" it as the `future` result. See my other comment in getResourceCountsByType().

Although, I'd actually consider getting rid of this method and using `getSubscribedResourcesMetadataSnapshot()` instead. It looks like it has access to all the data needed here and avoids polluting XdsClientImpl with the very specific form of data you need. In particular, "nacked_but_cached" seems highly specific to this one consumer of the data.

(You can contrast this method with `reportServerConnections()`, which works well as a callback as it just loops over the existing data. Although the data it returns is also small, so it could have been fine to copy it.)

DNVindhya replied:
`getSubscribedResourcesMetadataSnapshot()` is missing `subscriber.data` which tells us if resource is cached and used in determining `grpc.xds.cache_state` attribute value. 
As discussed, I will try adding `subcriber.data` value as a new field in ResourceMetadata and after that I will move the processing / formatting of values to `XdsClientMetricReporterImpl`.

ejona86 replied:
You might just add a `boolean` for whether there is data cached. That way it does stay "metadata."

DNVindhya replied:
Moved resource count by cache state metric aggregating and reporting to `XdsClientmetricReporterImpl". Added a boolean `isCached` to `ResourceMetadata`. PTAL

ejona86 commented on xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java:
```diff
@@ -504,9 +533,69 @@ private <T extends ResourceUpdate> void handleResourceUpdate(
+
+  /** Calculates number of resources by ResourceType and ResourceSubscriber.metadata.status. */
+  Map<XdsResourceType<?>, Map<String, Long>> getResourceCountsByType() {
+    Map<XdsResourceType<?>, Map<String, Long>> resourceCountsByType = new HashMap<>();
```
This map is only written to in this method, so it isn't actually necessary. If `ResourceCallback` was an argument to this method:
```java
// instead of:
      resourceCountsByType.put(resourceType, resourceCountsByState);
// you could do:
      resourceCountsByState.forEach((cacheState, count) ->
          callback.reportResourceCountGauge(count, cacheState, resourceType.typeUrl(), target));
```

There's not many XdsResourceType nor cache_state, so creating the full map isn't that much memory. Most of the cost right now is `toLowerCase()` and `Long` object creation for each entry. And optimizing those has no real impact to the memory cost of the data structure you return.

Although it looks like this is going to have to be reorganized when `authority` is added, that still won't add much additional memory.

But to really benefit from the callback, we probably shouldn't do any aggregation within XdsClientImpl. Instead, call the callback with the raw data and have it aggregate as it needs. That means passing `ResourceMetadataStatus` and `isResourceCached` separately and letting it determine the string to use. That makes the implementation here less sensitive to the whims of the metric reporter.

DNVindhya replied:
I initially separated to into multiple methods for testing the logic of aggregating resource counts separately, but instead tested using callback itself. Removed `getResourceCountsByType()` and `reportResourceCountsToCallback()`. 

> Instead, call the callback with the raw data and have it aggregate as it needs.
Will make this change after populating `subscriber.data` value in `ResourceMetadata`.

ejona86 commented on xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java:
```diff
@@ -109,7 +112,9 @@ public XdsClientImpl(
       MessagePrettyPrinter messagePrinter,
-      Object securityConfig) {
+      Object securityConfig,
+      String target,
```
Don't pass `target` into XdsClientImpl. It isn't using it; it is just echoing it right back out. Instead, `target` should be remembered by `XdsClientMetricReporterImpl` and "injected" while reporting.

DNVindhya replied:
Done

ejona86 commented on xds/src/main/java/io/grpc/xds/SharedXdsClientPoolProvider.java:
```diff
@@ -111,19 +115,17 @@ public ImmutableList<String> getTargets() {
   @ThreadSafe
   @VisibleForTesting
-  static class RefCountedXdsClientObjectPool implements ObjectPool<XdsClient> {
+  protected class RefCountedXdsClientObjectPool implements ObjectPool<XdsClient> {
```
`protected` is weird to see here, as `SharedXdsClientPoolProvider` is `final`. Revert back to package-private? (It doesn't actually change anything, but just confusing.)

ejona86 commented on xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:
```diff
@@ -0,0 +1,174 @@
+ * XdsClientMetricReporter implementation.
+ */
+@Internal
+public class XdsClientMetricReporterImpl implements XdsClientMetricReporter {
```
Why is this public? Make it package-private?

ejona86 commented on xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:
```diff
@@ -0,0 +1,174 @@
+
+  void close() {
+    if (gaugeRegistration != null) {
+      gaugeRegistration.close();
```
Since the reporter is reused, let's also set `gaugeRegistration = null` just after this, so it'll be a clearer when debugging.

ejona86 commented on xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:
```diff
@@ -0,0 +1,174 @@
+    this.gaugeRegistration = metricRecorder.registerBatchCallback(new BatchCallback() {
+      @Override
+      public void accept(BatchRecorder recorder) {
+        reportCallbackMetrics(recorder);
```
Since you're making an anonymous class anyway, you can also pass XdsClient here. Less state.

ejona86 commented on xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:
```diff
@@ -0,0 +1,174 @@
+
+  void setXdsClient(XdsClient client) {
+    this.xdsClient = client;
+    // register gauge here
```
Since the reporter is reused, I suggest at least adding `assert gaugeRegistration == null` to catch lifetime bugs. (Which would require setting it to null in close().)

ejona86 commented on xds/src/main/java/io/grpc/xds/client/ControlPlaneClient.java:
```diff
@@ -224,6 +225,25 @@ void readyHandler() {
+   * upon receiving the first response on a new ADS stream.
+   */
+  // Must be synchronized
+  boolean hasWorkingAdsStream() {
```
I think this method should just be `return !streamClosedNoResponse`. `shutdown` shouldn't matter, since this method should not be called when shutdown. `if (adsStream == null) return true` is the same as `if (adsStream == null) return !streamClosedNoResponse` because of earlier checks. And `responseReceived == !streamClosedNoResponse` when `adsStream.closed == true`. When `adsStream.closed == false`, it seems `!streamClosedNoResponse` is more correct than `responseReceived`, as we don't want to incorrectly say it is working.

DNVindhya commented on xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java:
```diff
@@ -394,7 +394,27 @@ public ControlPlaneClient getOrCreateControlPlaneClient(ServerInfo serverInfo) {
         xdsTransport,
         serverInfo,
         bootstrapInfo.node(),
-        this,
```
Counter of xDS servers going from healthy to unhealthy  metric (`grpc.xds_client.server_failure`) is reported as part of `handleStreamClosed` and the metric needs `grpc.xds.server` as an attribute value. 
Now with the help of anonymous  class we are able to create a `XdsResponseHandler` instance for every ControlPlaneClient and provide ControlPlaneClient's `ServerInfo` to `XdsClientImpl.this.handleStreamClosed`.

DNVindhya commented on xds/src/main/java/io/grpc/xds/client/XdsClient.java:
```diff
@@ -390,16 +401,14 @@ public Map<Bootstrapper.ServerInfo, LoadReportClient> getServerLrsClientMap() {
   /**
-   * Reports whether xDS client has a working ADS stream to xDS server. Reporting is done through
-   * {@link CallbackMetricReporter}.
+   * Reports whether xDS client has a working ADS stream to xDS server.
```
Added quotes around working.

ejona86 commented on xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:
```diff
@@ -0,0 +1,219 @@
+  private SettableFuture<Void> computeAndReportResourceCounts(
+      Map<XdsResourceType<?>, Map<String, ResourceMetadata>> metadataByType,
+      MetricReporterCallback callback) {
+    SettableFuture<Void> future = SettableFuture.create();
```
Delete? The future doesn't seem to be doing anything because this runs in-line.

ejona86 commented on xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:
```diff
@@ -0,0 +1,219 @@
+
+  private static String cacheStateFromResourceStatus(ResourceMetadataStatus metadataStatus,
+      boolean isResourceCached) {
+    String status = metadataStatus.toString().toLowerCase(Locale.ROOT);
```
Instead of toLower, which is going to allocate every item, can we just have a `switch` here that returns string literals?

ejona86 commented on xds/src/main/java/io/grpc/xds/client/XdsClient.java:
```diff
@@ -378,6 +386,27 @@ public Map<Bootstrapper.ServerInfo, LoadReportClient> getServerLrsClientMap() {
+   * @see <a href="https://github.com/grpc/proposal/blob/master/A78-grpc-metrics-wrr-pf-xds.md#xdsclient">
+   *   A78-grpc-metrics-wrr-pf-xds.md</a>
+   */
+  public SettableFuture<Void> reportServerConnections(ServerConnectionCallback callback) {
```
The return type should be `ListenableFuture<Void>` (or `Future<Void>`, if we don't want/care about notifications). `SettableFuture` should always be an implementation detail.

DNVindhya replied:
TIL, updated to return `Future<Void>` instead.

DNVindhya commented on xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:
```diff
@@ -41,7 +50,7 @@
  * XdsClientMetricReporter implementation.
  */
 @Internal
```
Removed

ejona86 commented on xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:
```diff
@@ -203,7 +200,6 @@ static final class MetricReporterCallback implements ResourceCallback,
 
     // TODO(dnvindhya): include the "authority" label once xds.authority is available.
-    @Override
     public void reportResourceCountGauge(long resourceCount, String cacheState,
```
private or package-private? Or just inline it? It doesn't get any benefit from being here instead of the parent class.

DNVindhya replied:
Updated to package-private.

DNVindhya commented:
@ejona86 and @larry-safran thanks for both of your approvals. I have addressed all your comments. If you don't have any new/pending comments can I go ahead and merge?

larry-safran commented:
Once you have approvals from everyone, you can go ahead a merge.

On Mon, Nov 25, 2024 at 3:06 PM Vindhya Ningegowda ***@***.***>
wrote:

> @ejona86 <https://github.com/ejona86> and @larry-safran
> <https://github.com/larry-safran> thanks for both of your approvals. I
> have addressed all your comments. If you don't have any new/pending
> comments can I go ahead and merge?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc-java/pull/11661#issuecomment-2499214460>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZQMCXTO7DVWXV7XYA4NHNT2COUQFAVCNFSM6AAAAABRBI57B6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIOJZGIYTINBWGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
